### PR TITLE
[Drop-In UI] Updated Store to allow Action Dispatch from State observers.

### PR DESCRIPTION
**Pull request by tomaszrybakiewicz**
### Description
-  Updated Store to allow Action Dispatch from State observers.

### CODE REVIEW

**Review by Zayankovsky — APPROVED** [review]
This looks great. I confirmed locally that we can revert `CameraComponent` changes in #6449 now.

File: libnavui-app/src/main/java/com/mapbox/navigation/ui/app/internal/Store.kt [hunk 1]
```diff
@@ -77,11 +77,14 @@ open class Store {
             return
         }
 
-        isDispatching = true
         if (!intercept(action)) {
-            reduce(action)
+            isDispatching = true
```
Comment (kmadsen) [inline libnavui-app/src/main/java/com/mapbox/navigation/ui/app/internal/Store.kt:81]: Does this all mean this change can be reverted? https://github.com/mapbox/mapbox-navigation-android/pull/6449

Comment (tomaszrybakiewicz) [reply libnavui-app/src/main/java/com/mapbox/navigation/ui/app/internal/Store.kt:81]: Yes but as a separate PR. 
Also lets keep the `UIComponent.observe` method improvents.

Comment (kmadsen) [reply libnavui-app/src/main/java/com/mapbox/navigation/ui/app/internal/Store.kt:81]: Awesome!

File: libnavui-app/src/main/java/com/mapbox/navigation/ui/app/internal/Store.kt [hunk 2]
```diff
@@ -91,11 +94,10 @@ open class Store {
         return false
     }
 
-    private fun reduce(action: Action) {
-        reducers.forEach { reducer ->
-            _state.value = reducer.process(_state.value, action)
+    private fun reduce(state: State, action: Action): State =
```
Comment (kmadsen) [inline libnavui-app/src/main/java/com/mapbox/navigation/ui/app/internal/Store.kt:97]: Should the same thing be done for the `intercept` function?

Curious if there is a race condition with the `_state.value`

``` kotlin
    private fun intercept(state: State, action: Action): Boolean {
        middlewares.forEach {
            if (it.onDispatch(state, action)) return true
        }
        return false
    }
```

**Review by kmadsen — APPROVED** [review]
🚢

File: libnavui-app/src/main/java/com/mapbox/navigation/ui/app/internal/Store.kt (hunk 2 above)
Comment (tomaszrybakiewicz) [reply libnavui-app/src/main/java/com/mapbox/navigation/ui/app/internal/Store.kt:102]: I wouldn't worry about this. There might be situations where we would like to register a middleware that consumes an action and dispatches a different action instead.

**Comment by abhishek1508** [comment]
We should revert the changes introduced in https://github.com/mapbox/mapbox-navigation-android/pull/6449 since the fix in this PR is what we want. Any objections?

cc @tomaszrybakiewicz @Zayankovsky

**Comment by abhishek1508** [comment]
> > We should revert the changes introduced in #6449 since the fix in this PR is what we want. Any objections?
> > 
> > cc @tomaszrybakiewicz @Zayankovsky
> 
> I wouldn't worry about it. Changes introduced in #6449 just change a Dispatcher, which is absolutely fine. Reverting it won't impact performance. So for now.... leave it.

Yeah but why do it if it's not needed after these changes. This makes me think what happens if we comment the changes introduced in #6449. Would the issue that it fixed still work with the changes in this PR?

**Comment by tomaszrybakiewicz** [comment]
> > > We should revert the changes introduced in #6449 since the fix in this PR is what we want. Any objections?
> > > cc @tomaszrybakiewicz @Zayankovsky
> > 
> > 
> > I wouldn't worry about it. Changes introduced in #6449 just change a Dispatcher, which is absolutely fine. Reverting it won't impact performance. So for now.... leave it.
> 
> Yeah but why do it if it's not needed after these changes. This makes me think what happens if we comment the changes introduced in #6449. Would the issue that it fixed still work with the changes in this PR?

I see no reason why it wouldn't. Either way, your call. But, let's apply any additional changes in a separate PR if needed.

**Review by abhishek1508 — APPROVED** [review]

**Pull request merged**